src/coreclr/vm/wasm/helpers.cpp

```diff
@@ -645,30 +668,54 @@ namespace
 
     typedef MapSHash<ULONG, const ReverseThunkMapValue*> HashToReverseThunkHash;
     HashToReverseThunkHash* reverseThunkCache = nullptr;
+    HashToReverseThunkHash* reverseThunkFallbackCache = nullptr;
+
+    HashToReverseThunkHash* CreateReverseThunkHashTable(bool fallback)
+    {
+        HashToReverseThunkHash* newTable = new HashToReverseThunkHash();
+        newTable->Reallocate(g_ReverseThunksCount * HashToReverseThunkHash::s_density_factor_denominator / HashToReverseThunkHash::s_density_factor_numerator + 1);
+        for (size_t i = 0; i < g_ReverseThunksCount; i++)
+        {
+            newTable->Add(fallback ? g_ReverseThunks[i].fallbackKey : g_ReverseThunks[i].key, &g_ReverseThunks[i].value);
+        }
+
+        HashToReverseThunkHash **ppCache = fallback ? &reverseThunkFallbackCache : &reverseThunkCache;
+        if (InterlockedCompareExchangeT(ppCache, newTable, nullptr) != nullptr)
+        {
+            // Another thread won the race, discard ours
+            delete newTable;
+        }
+        return *ppCache;
+    }
 
     const ReverseThunkMapValue* LookupThunk(MethodDesc* pMD)
     {
         HashToReverseThunkHash* table = VolatileLoad(&reverseThunkCache);
         if (table == nullptr)
         {
-            HashToReverseThunkHash* newTable = new HashToReverseThunkHash();
-            newTable->Reallocate(g_ReverseThunksCount * HashToReverseThunkHash::s_density_factor_denominator / HashToReverseThunkHash::s_density_factor_numerator + 1);
-            for (size_t i = 0; i < g_ReverseThunksCount; i++)
-            {
-                newTable->Add(g_ReverseThunks[i].key, &g_ReverseThunks[i].value);
-            }
-
-            if (InterlockedCompareExchangeT(&reverseThunkCache, newTable, nullptr) != nullptr)
-            {
-                // Another thread won the race, discard ours
-                delete newTable;
-            }
-            table = reverseThunkCache;
+            table = CreateReverseThunkHashTable(false /* fallback */);
         }
 
         ULONG key = CreateKey(pMD);
 
+        // Try primary key, it is based on Assembly fully qualified name and method token
         const ReverseThunkMapValue* thunk;
+        if (table->Lookup(key, &thunk))
+        {
+            return thunk;
+        }
+
+        // Try fallback key, that is based on method properties and assembly name
+        // The fallback is used when the assembly is trimmed and the token and assembly fully qualified name
+        // may change.
+        table = VolatileLoad(&reverseThunkFallbackCache);
+        if (table == nullptr)
+        {
+            table = CreateReverseThunkHashTable(true /* fallback */);
+        }
+
+        key = CreateFallbackKey(pMD);
+
```

Comment:
What happens if we get a key collision? Is it mechanically prevented from happening somewhere?